fdb-record-layer-lucene/src/main/java/org/apache/lucene/codecs/lucene84/LuceneOptimizedPostingsReader.java

```diff
@@ -66,55 +67,66 @@ public final class LuceneOptimizedPostingsReader extends PostingsReaderBase {
 
     private static final long BASE_RAM_BYTES_USED = RamUsageEstimator.shallowSizeOfInstance(Lucene84PostingsReader.class);
 
-    private final IndexInput docIn;
-    private final IndexInput posIn;
-    private final IndexInput payIn;
+    private Supplier<IndexInput> docIn;
+    private boolean docInInitialized;
+    private Supplier<IndexInput> posIn;
+    private boolean posInInitialized;
+    private Supplier<IndexInput> payIn;
+    private boolean payInInitialized;
 
-    private final int version;
+    private Supplier<Integer> version;
 
     /**
-     * Sole constructor.
-     * @param state reader parameters
-     * @throws IOException if there is an error in the underlying storage
+     * Sole  constructor.
+     *
+     * @param state state
+     * @throws IOException exception
      */
-    @SuppressWarnings("PMD.CloseResource")
     public LuceneOptimizedPostingsReader(SegmentReadState state) throws IOException {
-        boolean success = false;
-        IndexInput docIn = null;
-        IndexInput posIn = null;
-        IndexInput payIn = null;
 
         // NOTE: these data files are too costly to verify checksum against all the bytes on open,
         // but for now we at least verify proper structure of the checksum footer: which looks
         // for FOOTER_MAGIC + algorithmID. This is cheap and can detect some forms of corruption
         // such as file truncation.
 
         String docName = IndexFileNames.segmentFileName(state.segmentInfo.name, state.segmentSuffix, Lucene84PostingsFormat.DOC_EXTENSION);
-        try {
-            docIn = state.directory.openInput(docName, state.context);
-            if (state.fieldInfos.hasProx()) {
-                String proxName = IndexFileNames.segmentFileName(state.segmentInfo.name, state.segmentSuffix, Lucene84PostingsFormat.POS_EXTENSION);
-                posIn = state.directory.openInput(proxName, state.context);
-                if (state.fieldInfos.hasPayloads() || state.fieldInfos.hasOffsets()) {
-                    String payName = IndexFileNames.segmentFileName(state.segmentInfo.name, state.segmentSuffix, Lucene84PostingsFormat.PAY_EXTENSION);
-                    payIn = state.directory.openInput(payName, state.context);
+        docIn = Suppliers.memoize(() -> {
+            try {
+                docInInitialized = true;
```

Comment:
This will set the initialized to `true` even if the call to `state.directory.openInput` fails with an exception, which would cause the call in line 190 to try the `get()` again, which may throw the exception again, which may skip the rest of the `close()` calls in lines 193 and 196.